runtime/src/bank.rs

```diff
@@ -2948,6 +3022,85 @@ impl Bank {
         vote_with_stake_delegations_map
     }
 
+    /// Lookup vote account.
+    /// First, lookup in the *cached_vote_accounts. If not fund, look up directly in AccountsDB.
+    ///
+    /// * cached_vote_accounts: vote accounts cached in StakeCache
+    /// * vote_pubkey: vote account pubkey to lookup
+    ///
+    /// Returns vote account if found.
+    fn get_vote_account_with_cache(
+        &self,
+        cached_vote_accounts: &VoteAccounts,
+        vote_pubkey: &Pubkey,
+    ) -> Option<VoteAccount> {
+        let solana_vote_program: Pubkey = solana_vote_program::id();
+        cached_vote_accounts.get(vote_pubkey).cloned().or_else(|| {
+            self.get_account_with_fixed_root(vote_pubkey)
+                .and_then(|account| VoteAccount::try_from(account).ok())
+                .filter(|vote_account| vote_account.owner() == &solana_vote_program)
+        })
+    }
+
+    /// Calculates epoch reward points from stake/vote accounts with reward calculate parameter struct.
+    ///
+    /// * reward_calculate_params: reward calculate parameter struct - stake history, delegation and vote accounts.
+    /// * rewards: epoch rewards in lamports
+    /// * thread_pool: instance of thread_pool for parallel processing
+    /// * metrics: reward metrics
+    ///
+    /// Returns reward lamports and points for the epoch.
+    fn calculate_reward_points2(
+        &self,
+        reward_calculate_params: &EpochRewardCalculateParamInfo,
+        rewards: u64,
+        thread_pool: &ThreadPool,
+        metrics: &mut RewardsMetrics,
+    ) -> Option<PointValue> {
+        let EpochRewardCalculateParamInfo {
+            stake_history,
+            stake_delegations,
+            cached_vote_accounts,
+        } = reward_calculate_params;
+
+        let (points, measure) = measure!(thread_pool.install(|| {
+            stake_delegations
+                .par_iter()
+                .filter_map(|(_stake_pubkey, stake_account)| {
+                    let vote_pubkey = stake_account.delegation().voter_pubkey;
+
+                    self.get_vote_account_with_cache(cached_vote_accounts, &vote_pubkey)
+                        .and_then(|vote_account| match vote_account.vote_state().deref() {
```

Comment:
ugh, `VoteAccount::vote_state()`'s interface is pain.  gimme a sec and i'll put up a refactor so this can just be `vote_account.vote_state().cloned().ok()`

Reply:
https://github.com/solana-labs/solana/pull/31256

Reply:
that pr is in now